### DISCUSSION TITLE
PluginManager fails to update itself

### DISCUSSION THREAD

**Discussion by QuantumQuacken** [aljawaid/PluginManager#65] [Q&A] [answered]
I'm trying to solve an issue with PluginManager failing to update itself from v4.2 to v4.5. When I press the update button, I get the error on the top of the page `You don't have the permission to remove this plugin.`. Otherwise, updating for all the other plugins works flawlessly. 
Checking the logs on the server, I see a HTTP GET with a 302 status and this PHP notice appear when pressing the update button:
```
2023/06/14 16:20:47 [error] 59830#59830: *65602 FastCGI sent in stderr: "PHP message: PHP Notice: 
tempnam(): file created in the system's temporary directory in /var/www/REDACTED/kanboard-1.2.30/app/Core/Plugin/Installer.php on line 103" while reading response header from upstream, client: REDACTED, server: REDACTED,
request: "GET /?controller=PluginController&action=update&archive_url=https%253A%252F%252Fgithub.com%252Faljawaid%252FPluginManager%252Freleases%252Fdownload%252Fv4.5%252FPluginManager-4.5.zip&csrf_token=REDACTED HTTP/2.0",
upstream: "fastcgi://unix:/var/run/php/php8.1-fpm-kanboard.sock:", host: "REDACTED"
```
the source of the notice is this [line of code](https://github.com/kanboard/kanboard/blob/main/app/Core/Plugin/Installer.php#L103)
```php
$archiveFile = tempnam(ini_get('upload_tmp_dir') ? ini_get('upload_tmp_dir') : sys_get_temp_dir(), 'kb_plugin');
```
and hints at the fact that I haven't set up an upload_tmp_dir, which instead I did, so I fiddled with the `php.ini` configs for about an hour, trying different combinations, but nothing different happened.
Also, this notice could be totally unrelated and misleading for PM's updating issue.
All the folder permissions seem to be fine, but maybe I am missing something?

What should I check to gather extra details about this issue? My feeling is that I misconfigured something, but I'm not sure what could be it at this point.

Thank you in advance for any help provided.

**Comment by aljawaid** [comment]
Hello... how did you first install PM?

**Reply by aljawaid** [reply]
I don't think this is related to PM either.  Sounds like a folder permissions issue.

I have had similar problems before, allow me to explain.

I have uploaded the files to my Kanboard using 'root' instead of 'user' (www-user)... when I try to use KB to uninstall or upgrade, I get the same permission error.  Usually, for me, it is because I uploaded my plugin to test as 'root' and then forgot to change the folder permissions for the plugin.

Also, my common mistake is to upload the .git related files and folders... I think they may cause problems if deleting without 'root' too.

In KB, the `.zip` version of the plugin does not contain any .git files/folders. Also, KB will install as a 'user' and not as 'root'

Hope that all makes sense.  

Let me know how you get on.

**Reply by aljawaid** [reply, marked as answer]
You can also try with https://github.com/aljawaid/KanboardSupport to see if anything comes up in them checks.

This is my test KB through KanboardSupport...

![image](https://github.com/aljawaid/PluginManager/assets/10233708/b1a4b7ba-03f4-426a-8a7f-b3c69d6b1afe)

Cross-check your folder permission with the screenshot or install the plugin.

**Reply by QuantumQuacken** [reply]
Hello!
The support plugin is amazing. I was losing my mind checking which of all the required paths had the correct owner and permissions, but this plugin definitely sped up this process.

Yes, it was a folder permission issue. It seems that I had to do `chown -R www-data:www-data kanboardDir/`, because the data, files and plugins directories (the ones shown in your screenshot) were all already configured correctly.
After running that command, the update went through without any other issue.

Thank you so much again!

**Reply by aljawaid** [reply]
> The support plugin is amazing. I was losing my mind checking which of all the required paths had the correct owner and permissions, but this plugin definitely sped up this process.

- glad you like it, its not that popular though but I put a lot of time into that.  I had the same struggles as you and it made upgrading such a pain and time consuming.  To be fair, all my plugins are a result of my struggles with upgrading KB and keeping my changes. Really glad it worked out for you.
Remember to star the repositories you use :)

**Comment by QuantumQuacken** [comment]
> I had the same struggles as you and it made upgrading such a pain and time consuming.  To be fair, all my plugins are a result of my struggles with upgrading KB and keeping my changes.

Well, isn't that one of the main reasons why people start coding stuff and uploading it on GitHub? :P
> Remember to star the repositories you use :)

First thing I did when I found PM last week (plus many of the others you have) :)

Thank you again!
